spring-boot/proxy/src/main/java/eu/coatrack/proxy/security/LocalApiKeyManager.java

```diff
@@ -0,0 +1,129 @@
+
+    public LocalApiKeyManager(
+            ApiKeyFetcher apiKeyFetcher,
+            @Value("${number-of-minutes-the-gateway-shall-work-without-connection-to-admin}") long minutes) {
+        this.apiKeyFetcher = apiKeyFetcher;
+        this.numberOfMinutesTheGatewayShallWorkWithoutConnectionToAdmin = minutes;
+    }
+
+    public ApiKey getApiKeyEntityByApiKeyValue(String apiKeyValue) {
+        log.debug("Trying to get the API key with the value {} from the local list.", apiKeyValue);
+
+        if (apiKeyValue == null)
+            return null;
+
+        Optional<ApiKey> optionalApiKey = localApiKeyList.stream().filter(
+                apiKeyFromLocalList -> apiKeyFromLocalList.getKeyValue().equals(apiKeyValue)).findFirst();
+        return optionalApiKey.orElse(null);
+    }
+
+    public boolean wasLatestUpdateOfLocalApiKeyListWithinDeadline() {
+        return LocalDateTime.now().isBefore(deadline);
+    }
+
+    @Async
+    @PostConstruct
+    @Scheduled(fixedRateString = "${local-api-key-list-update-interval-in-millis}")
+    public void updateLocalApiKeyList() {
+        log.debug("Trying to update the local API key list by contacting CoatRack admin.");
+
+        List<ApiKey> fetchedApiKeyList;
+        try {
+            fetchedApiKeyList = apiKeyFetcher.requestLatestApiKeyListFromAdmin();
+            Assert.notNull(fetchedApiKeyList, "Fetched API key list must not be null.");
+        } catch (Exception e) {
+            displayGatewayModeSwitchesConsideringCurrentGatewayMode(GatewayMode.OFFLINE);
+            log.debug("Following error occurred: " + e);
+            return;
+        }
+
+        localApiKeyList = fetchedApiKeyList;
+        deadline = LocalDateTime.now().plusMinutes(numberOfMinutesTheGatewayShallWorkWithoutConnectionToAdmin);
+        displayGatewayModeSwitchesConsideringCurrentGatewayMode(GatewayMode.ONLINE);
+    }
+
+    private void displayGatewayModeSwitchesConsideringCurrentGatewayMode(GatewayMode modeOfCurrentUpdateInterval) {
+        if (isSwitchingToOnlineMode(modeOfCurrentUpdateInterval)){
+            log.info("Connection to the CoatRack admin server could be established. Switching to online mode.");
+            lastModeDisplayedInLog = GatewayMode.ONLINE;
+        }
+        else if (isSwitchingToOfflineMode(modeOfCurrentUpdateInterval)){
+            log.info("There is a problem with CoatRack admin. Switching to offline mode.");
+            lastModeDisplayedInLog = GatewayMode.OFFLINE;
+        }
+        modeOfPreviousUpdateInterval = modeOfCurrentUpdateInterval;
+    }
+
+    private boolean isSwitchingToOnlineMode(GatewayMode modeOfCurrentUpdateInterval) {
+        return lastModeDisplayedInLog == GatewayMode.OFFLINE
+                && modeOfCurrentUpdateInterval == GatewayMode.ONLINE;
+    }
+
+    private boolean isSwitchingToOfflineMode(GatewayMode modeOfCurrentUpdateInterval) {
+        return modeOfPreviousUpdateInterval == GatewayMode.OFFLINE
```

Comment:
This feature was simplified and should be much more readable now. The second global variable was implemented because of some misunderstanding about the feature and is now removed.